data_structures/linked_list/singly_linked_list.py

```diff
@@ -72,7 +72,7 @@ def __len__(self) -> int:
         >>> len(linked_list)
         0
         """
-        return len(tuple(iter(self)))
+        return sum(1 for _ in self)
```

Comment:
The sum will call iter(self) inside,  so this code is still O(n).

Reply:
For the classical linked list, the best length is O(n).

PS. `collections.deque` is not a real linked list,  it's a wrapper around it, so it basically counts how many append/delete methods were called.    https://github.com/python/cpython/blob/main/Modules/_collectionsmodule.c#L196



https://wiki.python.org/moin/TimeComplexity

Reply:
Are you talking about "Time" complexity? This line would fix the "Space" complexity as it only fetches one item at a time and adds 1 to the final result as opposed to `tuple()` which loads all the items to the memory.

Reply:
Fun exercise... Go to https://pyodide.org/en/latest/console.html to get a current Python repl running on WASM.

Paste in the following code and hit return.
```
>>> from timeit import timeit
setup="from itertools import product; from string import ascii_letters"
timeit("sum(1 for _ in product(ascii_letters, repeat=4))", number=10, setup=setup)
timeit("len(tuple(product(ascii_letters, repeat=4)))", number=10, setup=setup)

5.0610000000000355
4.121999999999957
```
`sum()` is slower than `len()` for 7,311,616 items.
---
Refresh the webpage to clear out any clutter in memory...

Paste in the following code and hit return.
```
>>> from timeit import timeit
setup="from itertools import product; from string import ascii_letters"
timeit("sum(1 for _ in product(ascii_letters, repeat=5))", number=1, setup=setup)
timeit("len(tuple(product(ascii_letters, repeat=5)))", number=1, setup=setup)
```
26.686000000000035
Traceback (most recent call last):
  ...
MemoryError
```
`sum()` delivers an answer for 380,204,032 items but `len()` raises a `MemoryError`.